staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,386 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remotecommand
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/remotecommand"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/websocket"
+	"k8s.io/klog/v2"
+)
+
+const (
+	pingReadDeadline = 60 * time.Second
+	// should be less than pingReadDeadline
+	pingWriteDeadline = 10 * time.Second
+	PingPeriod        = 5 * time.Second
+)
+
+var (
+	_ Executor          = &wsStreamExecutor{}
+	_ streamCreator     = &wsStreamCreator{}
+	_ httpstream.Stream = &stream{}
+
+	streamType2streamID = map[string]byte{
+		v1.StreamTypeStdin:  remotecommand.StreamStdIn,
+		v1.StreamTypeStdout: remotecommand.StreamStdOut,
+		v1.StreamTypeStderr: remotecommand.StreamStdErr,
+		v1.StreamTypeError:  remotecommand.StreamErr,
+		v1.StreamTypeResize: remotecommand.StreamResize,
+	}
+)
+
+// wsStreamExecutor handles transporting standard shell streams over an httpstream connection.
+type wsStreamExecutor struct {
+	config    *restclient.Config
+	method    string
+	url       string
+	protocols []string
+}
+
+// NewWebSocketExecutor allows to execute commands via a WebSocket connection.
+func NewWebSocketExecutor(config *restclient.Config, method, url string) Executor {
```

Comment:
I'd suggest parity with the SPDY constructors by passing in an already-created transport, to avoid exposing the restclient.Config, and avoid reconstructing a transport on every request. Also, include an error in the constructor value even if it is unused currently

`func NewWebSocketExecutor(transport http.RoundTripper, method, url string) (Executor, error)`

Reply:
hmm, traced the spdy code further back and realized that the only in-tree use is calling the NewSPDYExecutor constructor which plumbs in restclient.config and constructs a single-use transport

given that, I guess this constructor is fine, but do the config → transport construction here and return an error if that fails, and store the rt/wsRT in wsStreamExecutor

Reply:
Done. The constructor requires the websocket `upgrader` `RoundTripper` as well.